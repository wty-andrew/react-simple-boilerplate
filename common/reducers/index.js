import { combineReducers } from 'redux'
import { connectRouter } from 'connected-react-router'

import globalReducer from './global/globalReducer'

const createRootReducer = history =>
  combineReducers({
    global: globalReducer,
    router: connectRouter(history),
  })

export default createRootReducer
