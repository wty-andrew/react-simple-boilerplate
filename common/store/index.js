import { createMemoryHistory } from 'history'
import { routerMiddleware } from 'connected-react-router'
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import { composeWithDevTools } from 'redux-devtools-extension/logOnlyInProduction'

import createRootReducer from '../reducers'

export const history = createMemoryHistory()
const middleware = [thunk, routerMiddleware(history)]

const options = {}
const composeEnhancers = composeWithDevTools(options)

export const initialState = {}

export const configureStore = (
  initialState = initialState,
  history = history
) => {
  const store = createStore(
    createRootReducer(history),
    initialState,
    composeEnhancers(applyMiddleware(...middleware))
  )

  if (module.hot) {
    module.hot.accept('../reducers', () => {
      store.replaceReducer(createRootReducer(history))
    })
  }
  return store
}
