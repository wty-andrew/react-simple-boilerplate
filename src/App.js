import React from 'react'
import { Route, Switch } from 'react-router-dom'
import { hot } from 'react-hot-loader'

import Home from './containers/Home'
import About from './containers/About'
import NotFoundPage from './containers/NotFoundPage'

import Header from './components/Header'

export const App = () => {
  return (
    <div>
      <Header />

      <Switch>
        <Route exact path="/" component={Home} />
        <Route path="/about" component={About} />
        <Route component={NotFoundPage} />
      </Switch>
    </div>
  )
}

export default hot(module)(App)
