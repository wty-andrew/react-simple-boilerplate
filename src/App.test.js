import React from 'react'
import { ConnectedRouter } from 'connected-react-router'
import { createMemoryHistory } from 'history'
import { Provider } from 'react-redux'
import { render, fireEvent } from '@testing-library/react'

import { App } from './App'
import { configureStore, initialState } from '../common/store'

const renderWithReduxAndRouter = (
  ui,
  {
    initialState,
    history = createMemoryHistory(),
    store = configureStore(initialState, history),
  } = {}
) => {
  return {
    ...render(
      <Provider store={store}>
        <ConnectedRouter history={history}>{ui}</ConnectedRouter>
      </Provider>
    ),
    history,
    store,
  }
}

describe('Full app rendering', () => {
  it('should renders correctly', () => {
    const { asFragment } = renderWithReduxAndRouter(<App />, { initialState })

    expect(asFragment()).toMatchSnapshot()
  })
})

describe('Full app navigation', () => {
  it('should start on home page', () => {
    const { store, getByText } = renderWithReduxAndRouter(<App />, {
      initialState,
    })

    const { router } = store.getState()
    expect(router.location.pathname).toBe('/')
    expect(getByText('You are home')).toBeInTheDocument()
  })

  it('should navigate to about page on link click', () => {
    const { store, getByText } = renderWithReduxAndRouter(<App />, {
      initialState,
    })

    fireEvent.click(getByText(/about/i))

    const { router } = store.getState()
    expect(router.location.pathname).toBe('/about')
    expect(getByText('You are on the about page')).toBeInTheDocument()
  })

  it('should navigate to not found page when url is invalid', () => {
    const { history, store, getByText } = renderWithReduxAndRouter(<App />, {
      initialState,
    })
    const invalidURL = '/some/bad/route'

    history.push(invalidURL)

    const { router } = store.getState()
    expect(router.location.pathname).toBe(invalidURL)
    expect(getByText('404 - Not Found')).toBeInTheDocument()
  })
})
