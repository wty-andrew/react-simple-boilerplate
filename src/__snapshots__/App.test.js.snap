// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render correctly 1`] = `
<div>
  <Header />
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      component={[Function]}
      path="/about"
    />
    <Route
      component={[Function]}
    />
  </Switch>
</div>
`;
